R/stat-sf.R

```diff
@@ -3,12 +3,49 @@
 #' @usage NULL
 #' @format NULL
 StatSf <- ggproto("StatSf", Stat,
-  compute_group = function(data, scales) {
-    bbox <- sf::st_bbox(data[[ geom_column(data) ]])
-    data$xmin <- bbox[["xmin"]]
-    data$xmax <- bbox[["xmax"]]
-    data$ymin <- bbox[["ymin"]]
-    data$ymax <- bbox[["ymax"]]
+  compute_layer = function(self, data, params, layout) {
+    # add coord to the params, so it can be forwarded to compute_group()
+    params$coord <- layout$coord
+    ggproto_parent(Stat, self)$compute_layer(data, params, layout)
+  },
+
+  compute_group = function(data, scales, coord) {
+    geometry_data <- data[[ geom_column(data) ]]
+    geometry_crs <- sf::st_crs(geometry_data)
+
+    bbox <- sf::st_bbox(geometry_data)
+
+    if (inherits(coord, "CoordSf")) {
+      # if the coord derives from CoordSf, then it
+      # needs to know about bounding boxes of geometry data
+      coord$record_bbox(
+        xmin = bbox[["xmin"]], xmax = bbox[["xmax"]],
+        ymin = bbox[["ymin"]], ymax = bbox[["ymax"]]
+      )
+
+      # register geometric center of each bbox, to give regular scales
+      # some indication of where shapes lie
+      bbox_trans <- sf_transform_xy(
+        list(
+          x = 0.5*(bbox[["xmin"]] + bbox[["xmax"]]),
+          y = 0.5*(bbox[["ymin"]] + bbox[["ymax"]])
+        ),
+        coord$get_default_crs(),
+        geometry_crs
+      )
```

Comment:
I want the regular scales to be as little involved as possible in determining the final extent, because for very non-linear or rotated coordinate systems taking the extremes in longitude and latitude and turning them into a box and projecting that will lead to an extent that is way too big. At the same time, I need to give the regular scales some numbers to work with, otherwise they'll set their ranges to `c(0, 1)` and that will mess things up even more. 

See the last two examples in this comment:
https://github.com/tidyverse/ggplot2/pull/3659#issuecomment-562818034
Specifically, in the top row of the very last set of plots, the final extent is a bit too large because the regular scales have seen too wide of an extent in longitude and latitude. This effect would be much more severe if we gave them the whole long-lat extent for every geometry.